src/common/vdom.ts

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  IDisposable
+} from 'phosphor/lib/core/disposable';
+
+import {
+  Message
+} from 'phosphor/lib/core/messaging';
+
+import {
+  clearSignalData, defineSignal, ISignal
+} from 'phosphor/lib/core/signaling';
+
+import {
+  Widget
+} from 'phosphor/lib/ui/widget';
+
+import {
+  render, VNode
+} from 'phosphor/lib/ui/vdom';
+
+
+/**
+ * An interface for a model to be used with vdom rendering.
+ */
+export
+interface IVDomModel extends IDisposable {
+  /**
+   * A signal emited when any model state changes.
+   */
+  stateChanged: ISignal<IVDomModel, void>;
+}
+
+
+/**
+ * Concrete implementation of IVDomModel.
+ */
+export class VDomModel implements IVDomModel {
+  /**
+   * Construct a new VDomModel.
+   */
+  constructor() {}
+
+  /**
+   * A signal emitted when any model state changes.
+   */
+  stateChanged: ISignal<IVDomModel, void>;
+
+  /**
+   * Dispose the model.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return
+    };
+    this._isDisposed = true;
+    clearSignalData(this);
+  }
+
+  /**
+   * Test whether the model is disposed.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  _isDisposed = false;
+}
+
+// Define the signals for the VDomModel class.
+defineSignal(VDomModel.prototype, 'stateChanged');
+
+
+/** 
+ * Phosphor widget that encodes best practices for VDOM based rendering.
+ */
+export
+abstract class VDomWidget<T extends IVDomModel> extends Widget {
+  /**
+   * Construct a var name = new type(arguments);VDomWidget.
+   */
+  constructor() {
+    super();
+  }
+
+  /**
+   * A signal emited when the model changes.
+   */
+  modelChanged: ISignal<VDomWidget<T>, void>;
+
+  /** 
+   * Set the model and fire changed signals.
+   */
+  set model(newValue: T) {
+    if (!newValue && !this._model || newValue === this._model) {
```

Comment:
I prefer to cast the `newValue` to `null` to ensure that `model` is always "something" or `null`, but never `undefined`.

````
newValue = newValue || null;
if (this._model === newValue) {
  return;
}
```